waspc/data/Generator/templates/react-app/src/auth/forms/Auth.tsx

```diff
@@ -194,59 +178,108 @@ const SubmitButton = styled('button', {
+  '&:disabled': {
+    opacity: 0.5,
+    cursor: 'not-allowed',
+    backgroundColor: '$gray400',
+    borderColor: '$gray400',
+    color: '$gray500',
+  },
   transitionTimingFunction: 'cubic-bezier(0.4, 0, 0.2, 1)',
   transitionDuration: '100ms'
 })
 
-const googleSignInUrl = `${config.apiUrl}{= googleSignInPath =}`
-const gitHubSignInUrl = `${config.apiUrl}{= gitHubSignInPath =}`
-
-// TODO(matija): introduce type for appearance
-const Auth = ({ isLogin, appearance, logo, socialLayout } :
-              { isLogin: boolean; logo: string; socialLayout: "horizontal" | "vertical" }) => {
-  const history = useHistory()
-
-  const [usernameFieldVal, setUsernameFieldVal] = useState('')
-  const [passwordFieldVal, setPasswordFieldVal] = useState('')
+const Message = styled('div', {
+  padding: '0.5rem 0.75rem',
+  borderRadius: '0.375rem',
+  marginTop: '1rem',
+})
 
-  const handleSubmit = async (event) => {
-    event.preventDefault()
-    try {
-      if (!isLogin) {
-        await signup({ username: usernameFieldVal, password: passwordFieldVal })
-      }
-      await login (usernameFieldVal, passwordFieldVal)
+const ErrorMessage = styled(Message, {
+  background: '$errorBackground',
+  color: '$errorText',
+})
 
-      setUsernameFieldVal('')
-      setPasswordFieldVal('')
+const SuccessMessage = styled(Message, {
+  background: '$successBackground',
+  color: '$successText',
+})
 
+{=# isGoogleAuthEnabled =}
+const googleSignInUrl = `${config.apiUrl}{= googleSignInPath =}`
+{=/ isGoogleAuthEnabled =}
+{=# isGitHubAuthEnabled =}
+const gitHubSignInUrl = `${config.apiUrl}{= gitHubSignInPath =}`
+{=/ isGitHubAuthEnabled =}
+
+function Auth ({ state, appearance, logo, socialLayout = 'horizontal' }: {
+    state: State;
+} & CustomizationOptions) {
+  const isLogin = state === "login";
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+  const [successMessage, setSuccessMessage] = useState<string | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  {=# isAnyPasswordBasedAuthEnabled =}
+  const history = useHistory();
+  const onErrorHandler = (error) => {
+    setErrorMessage(error.message)
+  };
+  {=/ isAnyPasswordBasedAuthEnabled =}
+  {=# isUsernameAndPasswordAuthEnabled =}
+  const { handleSubmit, usernameFieldVal, passwordFieldVal, setUsernameFieldVal, setPasswordFieldVal } = useUsernameAndPassword({
+    isLogin,
+    onError: onErrorHandler,
+    onSuccess() {
       // Redirect to configured page, defaults to /.
       history.push('{= onAuthSucceededRedirectTo =}')
-    } catch (err) {
-      console.log(err)
-      window.alert(errorMessage(err))
+    },
+  });
+  {=/ isUsernameAndPasswordAuthEnabled =}
+  {=# isEmailAuthEnabled =}
+  const { handleSubmit, emailFieldVal, passwordFieldVal, setEmailFieldVal, setPasswordFieldVal } = useEmail({
+    isLogin,
+    onError: onErrorHandler,
+    showEmailVerificationPending() {
+      setSuccessMessage('Check your email for a confirmation link.')
```

Comment:
I'd enrich the message a bit, sth like "You've signed up successfully! Check your email ..."

Reply:
I'll add it 👍